dashboard/app/models/pd/workshop.rb

```diff
@@ -185,19 +185,36 @@ def self.send_reminder_for_upcoming_in_days(days)
     end
   end
 
-  def self.send_exit_surveys
-    # TODO: handle teachers who show up, join the section, and are marked attended without enrolling.
-    end_in_days(0).each do |workshop|
-      workshop.enrollments.each do |enrollment|
-        teacher = User.find_by(email: enrollment.email, user_type: User::TYPE_TEACHER)
-        Pd::WorkshopMailer.exit_survey(workshop, teacher, enrollment).deliver_now
-      end
-    end
-  end
-
   def self.send_automated_emails
     send_reminder_for_upcoming_in_days(3)
     send_reminder_for_upcoming_in_days(10)
-    send_exit_surveys
+  end
+
+  def self.process_closed_workshop_async(id)
+    workshop = Pd::Workshop.find(id)
+    raise "Unexpected workshop state #{workshop.state}." unless workshop.state == STATE_ENDED
+
+    workshop.send_exit_surveys
+  end
+
+  def send_exit_surveys
+    # Update enrollments with resolved users.
+    self.enrollments.each do |enrollment|
+      enrollment.update!(user: enrollment.resolve_user) unless enrollment.user
+    end
+
+    Pd::Enrollment.create_for_unenrolled_attendees(self)
+
+    # Send the emails
+    self.enrollments.reload.each do |enrollment|
+      next unless enrollment.user
```

Comment:
What do you want to do here if a user is a teacher without an email? Is this worthy of an error / warning / honeybadger error?